Assign Population Improvements

itanasi opened the pull request:
Solves #6144 
With a lot of help from @SomeTroglodyte 
- [x] Make Tile Lock matter for reassignPopulation()
- [x] Reassign on new trigger at start of NextTurn. Population Growth, Improvement completed/pillaged, new Tech
- Make Unassigned Population text also reassign population
- [x] Add Reset Tiles button for quick tile unlocks and assignment
- [x] Add Manual vs Auto Specialists functionality and controls
- Auto switch to Manual when adding/removing Specialists
- Click on text to toggle back to Manual. Need to work on making that clear
- [x] Update Locked icon
- [x] Change manual population override to Lock Tile on first click
- This would match Civ5. But do people really want that?
- [x] Focus Production UI and Equations (stretch goal)
- [x] Avoid Growth (stretch)
- [x] Translation templates

Example:
![image](https://user-images.githubusercontent.com/44038014/168734251-779faa0d-8cd1-44ed-ba81-2f2de763e576.png)

TODOS for future contributions:
- Make ScrollPane to support smaller resolutions, or ExpanderTab
- AI usage of Focuses
- More speed improvements
- JSON the weights of Stats for mod support
- Great Person Focus

itanasi commented:
Could use feedback on UI text color choices
![image](https://user-images.githubusercontent.com/44038014/166091091-96bdfacf-91b1-4320-9219-8a0b1ff4ce03.png)
![image](https://user-images.githubusercontent.com/44038014/166091098-0ff97252-bcf6-4e69-b69b-ce8a8a2a5e83.png)

itanasi commented:
Changed to look more like a button
![image](https://user-images.githubusercontent.com/44038014/166091633-cdb348a9-fd27-4fa6-a629-400079ef0638.png)

SomeTroglodyte commented:
Nice. I always thought _some day_ I'd try to do something akin to Civ IV's governors... Your "stretch goal" seems to go in that direction. Cheers!

> feedback

- "Copy RGB" sounds like bad practice. We already have too much mix of visuals living in an incomplete skin and hardcoded (guilty myself)... You _can_ extract any skin element easily in code wherever you need it. `BaseScreen.skin.get(TextButton.TextButtonStyle::class.java).fontColor` or somesuch.
- I'd look to get a RoundedEdgeRectangle in there so the "toggle button" is unmistakable - actually _make it_ a TextButton then tweak it to shrink the space needed and make it less obtrusive, less padding, maybe make the RER color.a=0.5f...
- "Reassign on new trigger. Population Growth?" -> Improvement finished or pillaged...
- "Reset Tiles" in that location maybe mistakable to mean "un-own and re-acquire the numer of tiles I'm due" - like an RPG respec. "Reassign worked tuiles[^1]"? "Reset workers"?
- "Make Unassigned Population text also" - an UI label with a functionality is easily missed, unless you know of a common expectation... (guilty myself - e.g. some unmarked pedia links) In this case, it's close. Some _will_ guess without help and be extra happy about their discovery - we're now a <i>5</i>X game.
- https://kotlinlang.org/docs/coding-conventions.html#horizontal-whitespace
- `posFromEdge + max(tileTable.height, selectedConstructionTable.height)` that max is for convenience, actually only one of the  two is visible at any one time, right? I know, annoying that Gdx doesn't do a true "anchor" like windoze forms does.
- `.tr().toLabel(Color.WHITE)` two redundancies?

[^1]: Some regional "tuiles d'amandes" are sooooooo delicious... (tuiles: frog = shingles, like those on a roof)

itanasi commented:
Thanks for the feedback. Especially the tip on the skin color. Might be more clear as a really tight TextButton. Need to play with it more.

Updating on Improvements and Tech would be nice, but much more complex to keep track of. Figuring out which City it "belongs" to and then marking it Dirty. Tech would have to always fire before cities and mark all cities Dirty. So more a stretch goal for me right now given dev time.

Triggering the auto assignment is the true UX thing I'm noodling. Reset clears locks, but we also should have an easy way to apply Population without messing with the locks. One indirect way is actually to switch to Auto Specialists, but that's weird. In Civ5 you could click on the Focus options, but likely that will be a dropdown here to save screen space. The UI text is definitely a bit hidden, not to mention small. But haven't figured out a cleaner solution yet.

Using max() was the hack. Only one should be visible, but tracking which is on needs some more surgery.

itanasi commented:
Semi crazy idea: clicking on the production stats in the UI could expand and reveal the Production Focus options

itanasi commented:
Uh, can't figure out how to extract a Up button's background color. TextButton only has Color for Font, and Button has Drawable and relies on Batch.getColor()

SomeTroglodyte commented:
I know what you mean. Either "assume" the `button-c: {color: color}` part of Skin.json and do `skin.get("color", Color::class.java)` or... reflection? No, cast: `(skin.get(Button.ButtonStyle::class.java).up as NinePatchDrawable).patch.color`

I admit, a bit extreme.

itanasi commented:
The first one works, but is a hack. Would be nice if I could reference the up state and tie that back to the button-c in Skin.json

SomeTroglodyte commented:
Yes our ninepatches have very limited scalability, by design, especially down. For much smaller radii you have no other option than what the notification scroll does - wrap and scale.

itanasi commented:
OK, now updating on new Techs and Improvements built near city

itanasi commented:
Latest version with new Lock icon and buttons
![image](https://user-images.githubusercontent.com/44038014/166185236-3a1f9228-8f4e-4078-8f30-8df1dbfcdfe1.png)

itanasi commented:
Buttons! Highlighting! Changing Focuses!
![image](https://user-images.githubusercontent.com/44038014/166870857-e8c5752b-08ee-4233-8274-fe675a972fc9.png)

SomeTroglodyte commented:
Toying with the branch... At first glance, ***verrrry*** nice. Yes, indeeeeed. Which is why I can't help but think it could be nicer. :smile_cat: 

- Citizen Management as a button... Dunno. I'd have begun with a SelectBox hidden in an ExpanderTab up in the CityStatsTable. The reset button shouldn't be that permanent - inside said expander or CitizenManagementTable might be better .. ... ....?
- Can mods do something like "Choose a mayor: {Homer Simpson (default) / Bill Clinton (growth) / Isaac Newton (science) / Uncle Scrooge (Gold) / Ferdinand II of Aragon (Faith)}"? Label+SelectBox could do that.
- Testing with that Civ6 mod some issue foisted on me... :+1: no conflicts I can see. More specialists, so the Citizen management button can overlap CityStatsTable - still works.
- CitizenManagementTable: Icons left of buttons better? My opinion is near 51:49
- Citizen Management button - marginally irritating if clicked repeatedly - unless UI redesign -> hide if CitizenManagementTable visible?
- Reset Citizens button - marginally irritating if clicked repeatedly - cache result of first click for the lifetime of CityScreen, then compare and if equal disable button? Lots of effort for something only slightly cooler.

In case I forgot to say so: Well done.

- `if (surplusFood < 0){ foodYield * 8 } else { foodYield /= 2 }` - is that yours? I mean, * in that place does nothing (very curious there's neiter a warning nor an inspection for this). The whole function tastes icky... If it deals in Stats, why copy out the individual ones into separate vars in the first place? Stats was built for operations like Stats[focus] *= 3 (if focus _were_ a Stat, but one could easily set up mappable enums)
- `autoAssignPopulation` is shockingly slow. Pity `rankTileForCityWork` won't stay static for the duration of the function, which prevents some effective optimizations... One could save one rankTileForCityWork call and one rankSpecialist call per loop iteration, but that's not much.

Did I say this is ***neat***?

SomeTroglodyte commented:
<details><summary>Food for thought: What could be done with...</summary>

```kotlin
enum class CityFocus(val stat: Stat? = null) {
    NoFocus {
        override fun getStatMultiplier(stat: Stat) = 1f  // actually redundant, but that's two steps to see
    },
    FoodFocus(Stat.Food),
    ProductionFocus(Stat.Production),
    GoldFocus(Stat.Gold),
    ScienceFocus(Stat.Science),
    CultureFocus(Stat.Culture),
    GoldGrowthFocus {
        override fun getStatMultiplier(stat: Stat) = when(stat) {
            Stat.Production, Stat.Food -> 2f
            else -> 1f
        }
    },
    ProductionGrowthFocus {
        override fun getStatMultiplier(stat: Stat) = when(stat) {
            Stat.Production, Stat.Food -> 2f
            else -> 1f
        }
    },
    FaithFocus(Stat.Faith),
    GreatPersonFocus;

    open fun getStatMultiplier(stat: Stat) = when (this.stat) {
        stat -> 3f
        else -> 1f
    }

    fun applyWeightTo(stats: Stats) {
        for (stat in Stat.values()) {
            stats[stat] *= getStatMultiplier(stat)
        }
    }
}
```
</details>

SomeTroglodyte commented on core/src/com/unciv/ui/cityscreen/CitizenManagementTable.kt:
```diff
@@ -0,0 +1,73 @@
+        else
+            defaultCell.background = ImageGetter.getBackground(BaseScreen.skin.get("color", Color::class.java))
+        innerTable.add(defaultCell).growX().pad(3f).row()
+        for ((stat, _) in city.cityStats.currentCityStats) {
```
This iterates over the Stat's that the city has > 0f, not all of them. It effectively calls `Stats.asSequence()` through `Stats.iterator()`. Look it up - Stats.kt lines 178, 153. I'd iterate over `CityFocus.values()` and add a `disabled` property to the focus enum for a `continue`, or a function if we wanted to move the `isReligionEnabled` test there. The latter maybe not.

SomeTroglodyte commented on core/src/com/unciv/ui/cityscreen/CitizenManagementTable.kt:
```diff
@@ -0,0 +1,73 @@
+        for ((stat, _) in city.cityStats.currentCityStats) {
+            if (stat == Stat.Happiness) continue
+            if (stat == Stat.Faith && !city.civInfo.gameInfo.isReligionEnabled()) continue
+            val label = "${stat.name} Focus".tr().toLabel().addBorder(5f, Color.CLEAR)
```
Double `tr()` (toLabel does one) and for the complex governors maybe another property on the enum.

SomeTroglodyte commented on core/src/com/unciv/logic/city/CityInfo.kt:
```diff
@@ -638,21 +660,42 @@ class CityInfo {
         }
     }
 
+    fun isFocus(stat: Stat): Boolean {
```
The CityInfo regions are already broken, but no need to place this in "state-changing functions" - could be obsolete with my enum fields suggestion anyway

SomeTroglodyte commented on core/src/com/unciv/ui/cityscreen/CityScreen.kt:
```diff
@@ -81,10 +92,28 @@ class CityScreen(
 
         //stage.setDebugTableUnderMouse(true)
         stage.addActor(cityStatsTable)
+        val resetCitizensButton = "Reset Citizens".tr().toTextButton()
```
Double tr(), citizenManagementButton as well

itanasi commented:
Appreciate the feedback! My original idea was an expand tab, but did this version for quicker development.

Switching to enum class seems like a much better solution. I obviously was kludging a way to associate the Stat.<type> to the focuses.

The equations are very much not ideal yet and going to need more tweaks.

I don't think I'll put in the effort in this Pull for the AI to also try switching modes. That's what a few of the extra options are really for as I read the Civ5 GnK and Vox Populi equations.

Hopefully time later tonight

SomeTroglodyte commented:
I'll show you my 'ideas' as code in a few moments.

SomeTroglodyte commented:
... Nope, sorry, I mismanaged and the code is _gone_. All thanks to me not being able to debug at all without #6458, which I removed from my local copy of your branch the wrong way. Sorry. Irrecoverable except from memory, and I prefer preparing the weekend, RL calls.

Ideas lost:
- CityStatsTable stat icons click selects focus (with CityFocus.valueOf(Stat): CityFocus?)
- Tried .surroundWithCircle instead of square focus indicator - not bad, a bit light
- Moved skin get colors out of loops
- The spurious two Automation.rank* calls in that assign function - can go with `val (best, bestValue) = things.filter(...).associateWith { rank }.maxOrNull { it.value } ?: Map.Entry(null, 0f)` - except one needs a trick to get a Map.Entry compatible (null to 0f)
- Moved reset button into CitizenManagementTable
- Hide citizenManagementButtonHolder when citizenManagementVisible
- folded common code in rankStatsForCityWork and rankSpecialistForCityWork, partially into two CityFocus based functions
- both use a val Stats.clone() in place of those seven vars, finishing with .values.sum()

I'm sure you'll get how specifically from that badly worded brainstorm and the other comments.

itanasi commented:
Sounds like I have at least two evenings of work then haha

itanasi commented:
Got most of the ideas done. Ranking equations not quite done. But did muck a bit with adding base weights. May want those JSON-able eventually

I'd rather the Reset Tiles button sit outside the Citizen Table. Most people will be manually moving workers around, locking them, and may need quick access to unlock all tiles. I had considered a SelectBox initially. Difficultly is it would then take 2 actions to change Focus instead of 1 with the pseudo-radio button system now. And with the Citizen Management Button it's very obvious that this feature exists. Did hide when the Table is open though. ExpanderTab is on the TODO.

>     * The spurious two Automation.rank* calls in that assign function - can go with `val (best, bestValue) = things.filter(...).associateWith { rank }.maxOrNull { it.value } ?: Map.Entry(null, 0f)` - except one needs a trick to get a Map.Entry compatible (null to 0f)

Not sure about this one? I don't quite follow

AdityaMH commented:
Move "reset citizens" to inside border 🤔

SomeTroglodyte commented:
> don't quite follow

lemme see what I can remember. What are the problems in autoAssignPopulation and downstream?
- `rankStatsForCityWork` and `rankSpecialistForCityWork`have _lots_ of repeated code.
- Both use 7 discrete var's where a `val rankingYields = stats.clone()` as start would be more readable. That thing _is_ designed so property access is fast, a local var wouldn't be that much faster.
- Both have one `foodYield * 8` unused expression doing nothing. Was `*=` intended yes or no? Not sure. I guessed yes.
- `autoAssignPopulation` calls both first within a `.maxByOrNull` not keeping the function result after the comparison, calling it again for the winner instead. One could save the value calculated during the maxByOrNull loop:
```kotlin
// now
            val bestTile: TileInfo? = cityInfo.getTiles()
                    .filter { it.aerialDistanceTo(cityInfo.getCenterTile()) <= 3 }
                    .filterNot { it.providesYield() }
                    .maxByOrNull { Automation.rankTileForCityWork(it, cityInfo, foodWeight) }
            val valueBestTile = if (bestTile == null) 0f
            else Automation.rankTileForCityWork(bestTile, cityInfo, foodWeight)
// saving one rankTileForCityWork call
            val (bestTile, valueBestTile) = cityInfo.getTiles()
                .filter { it.aerialDistanceTo(cityInfo.getCenterTile()) <= 3 }
                .filterNot { it.providesYield() }
                .associateWith { Automation.rankTileForCityWork(it, cityInfo, foodWeight) }
                .maxByOrNull { it.value }
                ?: object : Map.Entry<TileInfo?, Float> {
                    override val key: TileInfo? = null
                    override val value = 0f
                } 
```
I know the fallback given to the Elvis is ugly but it's the best clear and kotlin-y way I found to always initialize both vals without crashing - if max was unsuccessful to null and 0f  respectively. And that's how to save one rankTileForCityWork call because the one used for maxing is stored directly. Dunno whether rankTileForCityWork is expensive enough to justify the ugliness, I'd guess yes.

Oh, and candidate tiles in this block are instantiated by ownership then filtered by distance, but instantiating by distance then filtering by ownership _might_ get cheaper in cities with many tiles if aerialDistanceTo is relatively expensive - Vector2 in HashSet<Vector2> might be far cheaper. Or not. Code like this:
```kotlin
            val tilesDistanceFirst = cityInfo.tileMap.getTilesInDistance(cityInfo.location, 3)
                .filter { it.position in cityInfo.tiles }
            val tilesOwnedFirst = cityInfo.getTiles()
                .filter { it.aerialDistanceTo(cityInfo.getCenterTile()) <= 3 }
            val (bestTile, valueBestTile) = (if (cityInfo.tiles.size > 18) tilesDistanceFirst else tilesOwnedFirst)
                .filterNot { it.providesYield() }
                .associateWith { Automation.rankTileForCityWork(it, cityInfo, foodWeight) }
                .maxByOrNull { it.value }
                ?: object : Map.Entry<TileInfo?, Float> {
                    override val key: TileInfo? = null
                    override val value = 0f
                }
```
... would need thorough measurements to determine if there's a worthwhile gain at all, and where the statistically best turnover point may be.

Too tired to actually inspect your progress, sorry. later.

SomeTroglodyte commented:
- Consider `fun statToFocus(stat: Stat) = values().firstOrNull { it.stat == stat } ?: NoFocus` - But I would probably call it CityFocus.safeValueOf(stat) because it's so close to what Enum.valueOf(String) does and the name has previous uses. Using CityFocus as 'namespace' works by placing the fun in the Enum's companion object.
- `for (stat in CityFocus.values()) {` - I would rename the var, it's no longer :Stat. Let's say `focus`...
- ... then `val label = "${stat.stat.name} Focus"` -> `focus.label` would IMHO be better. More code - another Enum property and filling it for each instance - but more flexible and easier to maintain
- ... then have NoFocus.label be "Default Focus" and see how much code you can save in CitizenManagementTable.update. Might need another property 'disabled' or 'hidden' instetad of stat==null
- `if (city.cityAIFocus == stat) cell.background =`... -> `cell.background = ImageGetter.getBackground(if (city.cityAIFocus == stat) colorSelected else colorButton)`? Interesting though why Studio does not pop the 'Lift assignment out of if' inspection - probably because the one-liner is 105 chars long.
- `val cell = Table()` if it's not a Cell<Actor> why call it cell? It's practically a Button - which in turn merits a comment why not  use Gdx TextButton (I know why but the next reader will wonder) - a comment which could be a `fun String.toSlimRectangularButton() = Table().apply { add(this.toLabel().addBorder(5f, Color.CLEAR); touchable = Touchable.enabled }` instead - name and whether to actually place it besides the existing .toTextButton and whether to give it an `action: ()->Unit` parameter and/or an `highlight: Boolean` one - up to debate.
- `surroundWithCircle` - I did something like `color = if(cityInfo.cityAIFocus.stat == stat) selected else Color.Clear` so no outer if or onClick duplication (or toggle behaviour) but it only looked OK without tweaking size - likely because I forgot the resizeActor parameter. Food for thought for readability, but your result is likely better.
- Losing foodWeight as parameter is interesting (as is commented out `food = (2.4f +...` or `if.. /=` inversion), I trust you'll do the balancing better than I would. There's two details remaining:  `yieldStats.food * 8` is still an unused expression, and you sure including happiness in rankStatsForCityWork is OK? The old code did it only for specialists, what reason could that have had...

yairm210 commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -19,43 +20,130 @@ import com.unciv.ui.victoryscreen.RankingType
+        
+        if (cityAIFocus != CityFocus.NoFocus && cityAIFocus != CityFocus.FoodFocus && cityAIFocus != CityFocus.ProductionGrowthFocus && cityAIFocus != CityFocus.GoldGrowthFocus){
+            // Focus on non-food/growth
+            if (surplusFood < 0){
+                // Starving, need Food, get to 0
+                yieldStats.food *= 8
+            } else {
+                yieldStats.food /= 2
+            }
+        } else {
+            // NoFocus or Food/Growth Focus. Target +2 Food Surplus
+            if (surplusFood < 2) {
+                yieldStats.food * 8
+            } else if (cityAIFocus != CityFocus.FoodFocus) {
+                yieldStats.food /= 2
+            }
```
There's too much indentation here for me, please remove the brackets on the inner ifelses

yairm210 commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -19,43 +20,130 @@ import com.unciv.ui.victoryscreen.RankingType
         } else {
-            if (stats.food <= 2 || city.civInfo.getHappiness() > 5) rank += stats.food * 1.2f * foodWeight // food get more value to keep city growing
-            else rank += (2.4f + (stats.food - 2) / 2) * foodWeight // 1.2 point for each food up to 2, from there on half a point
+            //if (stats.food <= 2 || city.civInfo.getHappiness() > 5) yieldStats.food *= 1.2f // food get more value to keep city growing
```
I assume you're going to remove dead comments later but just making sure :)

yairm210 commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -19,43 +20,130 @@ import com.unciv.ui.victoryscreen.RankingType
+
+        if (cityAIFocus != CityFocus.NoFocus && cityAIFocus != CityFocus.FoodFocus && cityAIFocus != CityFocus.ProductionGrowthFocus && cityAIFocus != CityFocus.GoldGrowthFocus){
+            // Starving, need Food, get to 0
+            if (surplusFood < 0){
+                yieldStats.food *= 8
+            } else {
+                yieldStats.food /= 2
+            }
+        } else {
+            // Target +2 Food Surplus
+            if (surplusFood < 2) {
+                yieldStats.food *= 8
+            } else if (cityAIFocus != CityFocus.FoodFocus) {
+                yieldStats.food /= 2
+            }
```
Same here, remove braces from inner if/elses

yairm210 commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -19,43 +20,130 @@ import com.unciv.ui.victoryscreen.RankingType
+            if (stats.food <= 2 || city.civInfo.getHappiness() > 5) yieldStats.food *= 1.2f // food get more value to keep city growing
+            else yieldStats.food = (2.4f + (stats.food - 2) / 2) // 1.2 point for each food up to 2, from there on half a point
+
+            if (!(city.civInfo.gold < 0 && city.civInfo.statsForNextTurn.gold <= 0))
```
if(!(a && b)) -> if(!a || !b) for better readability

yairm210 commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -19,43 +20,130 @@ import com.unciv.ui.victoryscreen.RankingType
+            if (!(city.civInfo.gold < 0 && city.civInfo.statsForNextTurn.gold <= 0))
+                yieldStats.gold /= 3 // 3 gold is worse than 2 production
+
+            if (!(city.tiles.size < 12 || city.civInfo.wantsToFocusOn(Victory.Focus.Culture)))
```
Same

yairm210 commented on core/src/com/unciv/ui/cityscreen/CitizenManagementTable.kt:
```diff
@@ -0,0 +1,64 @@
+        }
+        isVisible = true
+
+        val focusLabel = "Default Focus".toLabel().addBorder(5f, Color.CLEAR)
+        val defaultCell = Table()
+        defaultCell.add(focusLabel)
```
Instead of adding a 'clear border' you can defaultCell.add(focusLabel).pad(5f) - saves unnecessary objects

yairm210 commented on core/src/com/unciv/ui/cityscreen/CitizenManagementTable.kt:
```diff
@@ -0,0 +1,64 @@
+        defaultCell.add(focusLabel)
+        defaultCell.touchable = Touchable.enabled
+        defaultCell.onClick { city.cityAIFocus = CityFocus.NoFocus; city.reassignPopulation(); cityScreen.update() }
+        if (city.cityAIFocus == CityFocus.NoFocus)
+            defaultCell.background = ImageGetter.getBackground(BaseScreen.skin.get("selection", Color::class.java))
+        else
+            defaultCell.background = ImageGetter.getBackground(BaseScreen.skin.get("color", Color::class.java))
```
How about backgroundName = if/else; defaultCell.background = ImageGetter...backgroundName

yairm210 commented on core/src/com/unciv/ui/cityscreen/CityStatsTable.kt:
```diff
@@ -35,9 +37,25 @@ class CityStatsTable(val cityScreen: CityScreen): Table() {
+                    cityInfo.cityAIFocus = statToFocus(stat)
+                    cityInfo.reassignPopulation(); cityScreen.update()
+                }
+                miniStatsTable.add(icon).size(20f).padRight(5f)
```
Since this happens anyway, remove from if/else

itanasi commented:
Equations consolidated and updated. Added Avoid Growth feature. All other advice done. I don't like my GPP hack, wish there was a better way to cache it

Haven't done ExpanderTab reorg

SomeTroglodyte commented:
> Haven't done ExpanderTab reorg

That was just an idea that might not even look good. Leave it for later. Go fix it up. But - I'll be away for almost a day for review.

itanasi commented:
I need to rethink through the weights a bit more.

The expected behavior is that when you Focus, you maximize that metric (without starving). So all weights should be such that x3 their weight they beat everything else.

However, when you incorporate GPP, Happiness, and the Food bonuses, very quickly Specialists can dominate as the best choice. Especially Merchants late game (+3 Gold, +2 Science is A LOT). Which doesn't feel quite right to me?

SomeTroglodyte commented:
As long as
a) if you don't use focus the choices aren't worse and
b) if you focus the associated metric is almost guaranteed to not drop

.. and I'd guess that is fulfilled.. Fine-tuning can come later. Heck, the UI with the original weighting (except the focus x3) would already be welcome.

itanasi commented:
So doing some profiling and this doubles the amount of processing time for autoAssignPopulation(). From my test case save, ~1sec to ~2sec

itanasi commented:
K, putting it up for review.

Future TODOs by the enterprising:
* Move into a ExpanderTab
* AI integration
* More efficient speedups
* JSON the weights

itanasi commented:
Cancel that. Thought of an idea to handle Food

itanasi commented:
Ok, quick sanity check to make sure I'm not doing something dumb.

How about at the start we calculate current Food Surplus based on however many Citizens are assigned. Save this value. Then as we assign Citizens increment Food Surplus by amount generated, and potentially extra for the Specialist Unique. It's not as bulletproof as doing a full recalc. Especially if there's bonuses (Artemis). But gets us close enough imo.

Downside is approximating quirks like Food Cost per Population and doubles some of the logic for Food calculations (terrible tech debt mine). But this prevents doubling the compute time.

itanasi commented:
It's not an AI aspect.

Adding this in is a decent performance hit because we recalculate all of the City Stats each time we assign a Population. This keeps things up to date and let's us make decisions based on how much Food we're generating so that we can hit >=0 Food (not starving).

So my fix would be instead of calculating all of the Stats, get them at the start then have a rough approximation of Food Generated by tracking the additional food from each Population assigned. Downside is duplicates some logic (Specialist Food bonus), and won't capture special +Food bonuses

SomeTroglodyte commented:
... can't really help at the moment, not enough resource governor assigned to Unciv ;)

xlenstra commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -89,21 +130,23 @@ object Automation {
-            }) // there is absolutely no reason for you to make water units on this body of water.
+                    (it.isCityCenter() && it.getOwner() != city.civInfo)
+                            || (it.militaryUnit != null && it.militaryUnit!!.civInfo != city.civInfo)
+                }) // there is absolutely no reason for you to make water units on this body of water.
```
Are all these extra indentations android studio auto-formatting? I personally don't like them tbh, and the kotlin style guide agrees with me on that. Its a style thing so this won't affect my approval regardless

itanasi replied:
Yeah, did an Android Studio auto format pass. Can clean up a bit

SomeTroglodyte replied:
I even think the defaults for that changed ArcticFox - Bumblebee - Chipmunk. But - we could make these settings part of the project under version control.

itanasi replied:
Or at least mention in the Wiki. Done on my editor and fixed

xlenstra commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -21,41 +21,82 @@ import com.unciv.ui.victoryscreen.RankingType
-            } else rank += stats.culture / 2
+                yieldStats.gold *= 2 // We have a global problem
+
+            if (city.tiles.size < 12 || city.civInfo.wantsToFocusOn(Focus.Culture))
```
You could look into also taking the `Victory.Focus.Science` and other stats into account

xlenstra replied:
Though that might be more something for AI and not for players

itanasi replied:
Done

xlenstra commented on core/src/com/unciv/ui/cityscreen/SpecialistAllocationTable.kt:
```diff
@@ -14,7 +15,18 @@ class SpecialistAllocationTable(val cityScreen: CityScreen): Table(BaseScreen.sk
+        // Color of "color" coming from Skin.json that's loaded into BaseScreen
+        // 5 columns: unassignButton, AllocationTable, assignButton, SeparatorVertical, SpecialistsStatsTabe
+        if(cityInfo.manualSpecialists) {
+            val manualSpecialists = "Manual Specialists".tr().toLabel().addBorder(5f, BaseScreen.skin.get("color", Color::class.java))
```
`.toLabel()` already has an internal `.tr()` call, so no need to do it twice

SomeTroglodyte replied:
Similar to the little hint [in my comment 17 days back](https://github.com/yairm210/Unciv/pull/6650#issuecomment-1113937648)

itanasi replied:
Yup

xlenstra commented on core/src/com/unciv/logic/city/PopulationManager.kt:
```diff
@@ -109,34 +111,63 @@ class PopulationManager {
+        cityInfo.cityStats.update()  // calculate current stats with current assignments
+        // save temporaries
+        cityInfo.currentSurplusFood = cityInfo.foodForNextTurn().toFloat()  // save temporary for calculations
+        cityInfo.currentGPPBonus = 0
+        for (unique in cityInfo.getMatchingUniques(UniqueType.GreatPersonPointPercentage))
+            if (cityInfo.matchesFilter(unique.params[1]))
+                cityInfo.currentGPPBonus += unique.params[0].toInt()
+        // Sweden UP
+        for (otherCiv in cityInfo.civInfo.getKnownCivs()) {
+            if (!cityInfo.civInfo.getDiplomacyManager(otherCiv).hasFlag(DiplomacyFlags.DeclarationOfFriendship))
+                continue
+
+            for (ourUnique in cityInfo.civInfo.getMatchingUniques(UniqueType.GreatPersonBoostWithFriendship))
+                cityInfo.currentGPPBonus += ourUnique.params[0].toInt()
+            for (theirUnique in otherCiv.getMatchingUniques(UniqueType.GreatPersonBoostWithFriendship))
+                cityInfo.currentGPPBonus += theirUnique.params[0].toInt()
+        }
```
Isn't this straight up copy + pasted from `CityInfo:493::getGreatPersonPointsForNextTurn()`? I'd rather extract it to a new method and call that in both places, if we at any point in the future want to change this code we would otherwise have to do it in two places (and probably forget)

itanasi replied:
Pulled out now

xlenstra commented on core/src/com/unciv/logic/city/PopulationManager.kt:
```diff
@@ -109,34 +111,63 @@ class PopulationManager {
+        var specialistFoodBonus = 0f
+        for (unique in cityInfo.getMatchingUniques(UniqueType.FoodConsumptionBySpecialists))
+            if (cityInfo.matchesFilter(unique.params[1]))
+                specialistFoodBonus = -(unique.params[0].toFloat() / 100f) * 2f // base 2 food per Pop, and need to inverse negative
```
We set this each time instead of it adding together?

itanasi replied:
Good catch. Should be cumulative multiplication, same as in CityStats. Fixed

xlenstra reviewed (commented):
Few nitpicky things, but without removing the code duplication for calculating great person points I can't approve this. I'll also test it once that is addressed to see how the system works in practice.

It also seems like the AI doesn't actually choose a focus. Is that intentional (I'm fine with that, this is already an improvement as is) or are you still planning something with that?

SomeTroglodyte commented:
> the AI being able to pick a Focus

And it shouldn't, it should simply optimize without prejudice. If the code here helps, fine. But that's better left for a later PR.

> I'll also test it

Just for reference, I tested the state of the branch about 11 days ago and it was stable, helpful and from what I could see not breaking the game's rules.

itanasi commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -21,41 +21,82 @@ import com.unciv.ui.victoryscreen.RankingType
+                    yieldStats.food /= 2
+            } else if (!city.avoidGrowth) {
+                // NoFocus or Food/Growth Focus. Target +2 Food Surplus
+                if (surplusFood < 2)
```
The point is to *8 until you hit 2 Surplus. That is, rank Food as the ultimate overriding need, until+2 Surplus. Then the weights carry it from there in a more balanced fashion. Or using Food Focus still gives a *3

itanasi commented on core/src/com/unciv/logic/city/CityInfo.kt:
```diff
@@ -302,6 +351,8 @@ class CityInfo {
         toReturn.isOriginalCapital = isOriginalCapital
         toReturn.flagsCountdown.putAll(flagsCountdown)
         toReturn.demandedResource = demandedResource
+        toReturn.updateCitizens = updateCitizens
+        toReturn.cityAIFocus = cityAIFocus
```
Good catch

itanasi commented on core/src/com/unciv/logic/automation/Automation.kt:
```diff
@@ -21,41 +21,82 @@ import com.unciv.ui.victoryscreen.RankingType
+        return rank
+    }
+
+    private fun rankStatsForCityWork(stats: Stats, city: CityInfo, specialist: Boolean = false): Float {
```
Done

itanasi commented on core/src/com/unciv/logic/city/CityInfo.kt:
```diff
@@ -34,6 +35,49 @@ enum class CityFlags {
+    CultureFocus("${Stat.Culture.name} Focus", true, Stat.Culture),
+    GoldGrowthFocus("Gold Growth Focus", false) {
+        override fun getStatMultiplier(stat: Stat) = when (stat) {
+            Stat.Production, Stat.Food -> 2f
```
Yup

itanasi commented on core/src/com/unciv/logic/city/CityInfo.kt:
```diff
@@ -585,7 +636,11 @@ class CityInfo {
-        if (isPuppet) reassignPopulation()
+        if (isPuppet) reassignAllPopulation()
+        if (updateCitizens){
+            reassignPopulation()
```
Good catch. Changed

itanasi commented on core/src/com/unciv/logic/city/PopulationManager.kt:
```diff
@@ -109,34 +111,63 @@ class PopulationManager {
+            } else if (bestJob != null) {
+                specialistAllocations.add(bestJob, 1)
+                cityInfo.currentSurplusFood -= specialistFoodBonus
+            }
```
Logic bug!

itanasi commented:
Hookay. All comments sorted

itanasi commented on core/src/com/unciv/logic/city/PopulationManager.kt:
```diff
@@ -109,34 +112,49 @@ class PopulationManager {
+                if (bestTile != null) {
                     cityInfo.workedTiles = cityInfo.workedTiles.withItem(bestTile.position)
-            } else if (bestJob != null) specialistAllocations.add(bestJob, 1)
+                    cityInfo.cityStats.currentCityStats[Stat.Food] += bestTile.getTileStats(cityInfo, cityInfo.civInfo)[Stat.Food]
```
It should run and clean up my temp use of it at the end when it runs the update(). I opted to just reuse this instead of copying to a Transient Stats variable in CityInfo.

The only stat that the Automation needs is Food. Replicating all of the stats if a lot of redundant code. I could run the update() after each Citizen is assigned, but that doubled the runtime

itanasi replied:
I can do the change later tonight

itanasi replied:
Done

SomeTroglodyte commented:
- Getting `Gold Growth Focus` and P G Focus into templates
- Getting the lock into the atlas

..Later? Those Foci - they weren't actually shown when last I played this. I'll have to give it another short spin.

> `Conversation 91 Commits 39 Checks 1 Files changed 15`

Getting that first counter to exactly :100: the moment of the merge will give an extra special reward by the github team - that's what you mean by ding ding ding? Or only if all four reach 100, thus 4 times the emoji?

SomeTroglodyte reviewed (approved):
Tests just fine. The atlas isn't critical as there's the old "Lock" in exactly the same place of the texture; and the un-templated strings indeed aren't visible.

itanasi commented:
> * Getting `Gold Growth Focus` and P G Focus into templates
> 
>     * Getting the lock into the atlas
> 
> 
> ..Later? Those Foci - they weren't actually shown when last I played this.

Figured the atlas would regenerate. Easy fix to push a compiled one.

As for the spare Foci, those are more for the AI. Too many options in the Table will make it waaaay too big for smaller resolutions (already pretty big, would like ScrollPane to work but couldn't quite get it last night).

itanasi commented:
I mean I figured the release process would recompile those assets. Otherwise sorta annoying to keep merge conflict fixing. Whatever

itanasi commented:
> > Otherwise sorta annoying to keep merge conflict fixing
> 
> You don't, you just do either `pick remote` or `pick ours` and just regenerate.

Correct, just did that. I just was assuming someone/something else was doing a quick regeneration as part of release

SomeTroglodyte commented:
Nobody volunteered to make that 100th discussion entry? Aw.

All right, here it is. Grace period expired, in this goes.
